git: Disable tests that fail on ZFS with formD normalization

mweinelt opened the pull request:
These two tests are regularly creating problems for my hydra instance, because its builders run on ZFS and that makes them fail consistently.

Closes: #185882

N.B.: If someone tells me how to collect verbose test logs I can submit them upstream for consideration.

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux with ZoL Root
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

mweinelt committed:
git: Disable tests that fail on ZFS with formD normalization

These two tests are regularly creating problems for my hydra instance,
because its builders run on ZFS and that makes them fail consistently.

The issue has something to do with unicode normalization. My pools have
formD normalization configured, that might be the culprit in this case.

Closes: #185882

winterqt commented:
> If someone tells me how to collect verbose test logs I can submit them upstream for consideration.

Per [the test suite's Makefile](https://github.com/git/git/blob/d3fa443f97e3a8d75b51341e2d5bac380b7422df/t/Makefile#L12), can you try setting `GIT_TEST_OPTS` to `--verbose --debug`?

mweinelt commented:
Thanks, once set it tells me to instead use `--verbose-log`.

<details><summary>t3910-mac-os-precompose</summary>
<pre>
 cat t3910-mac-os-precompose.*
1
Initialized empty Git repository in /build/git-2.37.3/t/trash directory.t3910-mac-os-precompose/.git/
checking prerequisite: UTF8_NFD_TO_NFC

mkdir -p "$TRASH_DIRECTORY/prereq-test-dir-UTF8_NFD_TO_NFC" &&
(
	cd "$TRASH_DIRECTORY/prereq-test-dir-UTF8_NFD_TO_NFC" &&
	# check whether FS converts nfd unicode to nfc
	auml=$(printf "\303\244")
	aumlcdiar=$(printf "\141\314\210")
	>"$auml" &&
	test -f "$aumlcdiar"

)
prerequisite UTF8_NFD_TO_NFC ok
expecting success of 3910.1 'detect if nfd needed': 
	precomposeunicode=$(git config core.precomposeunicode) &&
	test "$precomposeunicode" = true &&
	git config core.precomposeunicode true

not ok 1 - detect if nfd needed
#	
#		precomposeunicode=$(git config core.precomposeunicode) &&
#		test "$precomposeunicode" = true &&
#		git config core.precomposeunicode true
#	

expecting success of 3910.2 'setup': 
	>x &&
	git add x &&
	git commit -m "1st commit" &&
	git rm x &&
	git commit -m "rm x"

[main (root-commit) 354305e] 1st commit
 Author: A U Thor <author@example.com>
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 x
rm 'x'
[main e53e959] rm x
 Author: A U Thor <author@example.com>
 1 file changed, 0 insertions(+), 0 deletions(-)
 delete mode 100644 x

expecting success of 3910.3 'setup case mac': 
	git checkout -b mac_os

ok 2 - setup
Switched to a new branch 'mac_os'
ok 3 - setup case mac

expecting success of 3910.4 'git diff f.Adiar': 
	touch f.$Adiarnfc &&
	git add f.$Adiarnfc &&
	echo f.Adiarnfc >f.$Adiarnfc &&
	git diff f.$Adiarnfd >expect &&
	git diff f.$Adiarnfc >actual &&
	test_cmp expect actual &&
	git reset HEAD f.Adiarnfc &&
	rm f.$Adiarnfc expect actual

--- expect	2022-09-17 01:53:41.854607088 +0000
+++ actual	2022-09-17 01:53:41.879607345 +0000
@@ -0,0 +1,6 @@
+diff --git "a/f.\303\204" "b/f.\303\204"
+index e69de29..3875f1e 100644
+--- "a/f.\303\204"
++++ "b/f.\303\204"
+@@ -0,0 +1 @@
++f.Adiarnfc
not ok 4 - git diff f.Adiar
#	
#		touch f.$Adiarnfc &&
#		git add f.$Adiarnfc &&
#		echo f.Adiarnfc >f.$Adiarnfc &&
#		git diff f.$Adiarnfd >expect &&
#		git diff f.$Adiarnfc >actual &&
#		test_cmp expect actual &&
#		git reset HEAD f.Adiarnfc &&
#		rm f.$Adiarnfc expect actual
#	

expecting success of 3910.5 'git diff-files f.Adiar': 
	touch f.$Adiarnfc &&
	git add f.$Adiarnfc &&
	echo f.Adiarnfc >f.$Adiarnfc &&
	git diff-files f.$Adiarnfd >expect &&
	git diff-files f.$Adiarnfc >actual &&
	test_cmp expect actual &&
	git reset HEAD f.Adiarnfc &&
	rm f.$Adiarnfc expect actual


ok 5 - git diff-files f.Adiar
expecting success of 3910.6 'git diff-index f.Adiar': 
	touch f.$Adiarnfc &&
	git add f.$Adiarnfc &&
	echo f.Adiarnfc >f.$Adiarnfc &&
	git diff-index HEAD f.$Adiarnfd >expect &&
	git diff-index HEAD f.$Adiarnfc >actual &&
	test_cmp expect actual &&
	git reset HEAD f.Adiarnfc &&
	rm f.$Adiarnfc expect actual

--- expect	2022-09-17 01:53:42.041609009 +0000
+++ actual	2022-09-17 01:53:42.064609243 +0000
@@ -0,0 +1 @@
+:000000 100644 0000000000000000000000000000000000000000 0000000000000000000000000000000000000000 A	"f.\303\204"
not ok 6 - git diff-index f.Adiar
#	
#		touch f.$Adiarnfc &&
#		git add f.$Adiarnfc &&
#		echo f.Adiarnfc >f.$Adiarnfc &&
#		git diff-index HEAD f.$Adiarnfd >expect &&
#		git diff-index HEAD f.$Adiarnfc >actual &&
#		test_cmp expect actual &&
#		git reset HEAD f.Adiarnfc &&
#		rm f.$Adiarnfc expect actual
#	

expecting success of 3910.7 'add file Adiarnfc': 
	echo f.Adiarnfc >f.$Adiarnfc &&
	git add f.$Adiarnfc &&
	git commit -m "add f.$Adiarnfc"

[mac_os ac01922] add f.Ä
 Author: A U Thor <author@example.com>
 1 file changed, 1 insertion(+)
 create mode 100644 "f.\303\204"

ok 7 - add file Adiarnfc
expecting success of 3910.8 'git diff-tree f.Adiar': 
	echo f.Adiarnfc >>f.$Adiarnfc &&
	git diff-tree HEAD f.$Adiarnfd >expect &&
	git diff-tree HEAD f.$Adiarnfc >actual &&
	test_cmp expect actual &&
	git checkout f.$Adiarnfc &&
	rm expect actual

--- expect	2022-09-17 01:53:42.171610333 +0000
+++ actual	2022-09-17 01:53:42.194610567 +0000
@@ -0,0 +1,2 @@
+ac01922d5aefa1073e5095fc1960687a05bdde63
+:000000 100644 0000000000000000000000000000000000000000 3875f1e9c89cf09272fd0e203f52d31d2fd9f7e4 A	"f.\303\204"
not ok 8 - git diff-tree f.Adiar
#	
#		echo f.Adiarnfc >>f.$Adiarnfc &&
#		git diff-tree HEAD f.$Adiarnfd >expect &&
#		git diff-tree HEAD f.$Adiarnfc >actual &&
#		test_cmp expect actual &&
#		git checkout f.$Adiarnfc &&
#		rm expect actual
#	

expecting success of 3910.9 'stage file d.Adiarnfd/f.Adiarnfd': 
	mkdir d.$Adiarnfd &&
	echo d.$Adiarnfd/f.$Adiarnfd >d.$Adiarnfd/f.$Adiarnfd &&
	git stage d.$Adiarnfd/f.$Adiarnfd &&
	git commit -m "add d.$Adiarnfd/f.$Adiarnfd"

[mac_os 3e00b16] add d.Ä/f.Ä
 Author: A U Thor <author@example.com>
 1 file changed, 1 insertion(+)
 create mode 100644 "d.A\314\210/f.A\314\210"

expecting success of 3910.10 'add link Adiarnfc': 
	ln -s d.$Adiarnfd/f.$Adiarnfd l.$Adiarnfc &&
	git add l.$Adiarnfc &&
	git commit -m "add l.Adiarnfc"

ok 9 - stage file d.Adiarnfd/f.Adiarnfd
[mac_os 7ec02a6] add l.Adiarnfc
 Author: A U Thor <author@example.com>
 1 file changed, 1 insertion(+)
 create mode 120000 "l.\303\204"
ok 10 - add link Adiarnfc

expecting success of 3910.11 'git log f.Adiar': 
	git log f.$Adiarnfc > f.Adiarnfc.log &&
	git log f.$Adiarnfd > f.Adiarnfd.log &&
	test -s f.Adiarnfc.log &&
	test -s f.Adiarnfd.log &&
	test_cmp f.Adiarnfc.log f.Adiarnfd.log &&
	rm f.Adiarnfc.log f.Adiarnfd.log

not ok 11 - git log f.Adiar
#	
#		git log f.$Adiarnfc > f.Adiarnfc.log &&
#		git log f.$Adiarnfd > f.Adiarnfd.log &&
#		test -s f.Adiarnfc.log &&
#		test -s f.Adiarnfd.log &&
#		test_cmp f.Adiarnfc.log f.Adiarnfd.log &&
#		rm f.Adiarnfc.log f.Adiarnfd.log
#	

expecting success of 3910.12 'git lsfiles f.Adiar': 
	git ls-files f.$Adiarnfc > f.Adiarnfc.log &&
	git ls-files f.$Adiarnfd > f.Adiarnfd.log &&
	test -s f.Adiarnfc.log &&
	test -s f.Adiarnfd.log &&
	test_cmp f.Adiarnfc.log f.Adiarnfd.log &&
	rm f.Adiarnfc.log f.Adiarnfd.log

not ok 12 - git lsfiles f.Adiar
#	
#		git ls-files f.$Adiarnfc > f.Adiarnfc.log &&
#		git ls-files f.$Adiarnfd > f.Adiarnfd.log &&
#		test -s f.Adiarnfc.log &&
#		test -s f.Adiarnfd.log &&
#		test_cmp f.Adiarnfc.log f.Adiarnfd.log &&
#		rm f.Adiarnfc.log f.Adiarnfd.log
#	

expecting success of 3910.13 'git mv': 
	git mv f.$Adiarnfd f.$Odiarnfc &&
	git mv d.$Adiarnfd d.$Odiarnfc &&
	git mv l.$Adiarnfd l.$Odiarnfc &&
	git commit -m "mv Adiarnfd Odiarnfc"

fatal: not under version control, source=f.Ä, destination=f.Ö
not ok 13 - git mv
#	
#		git mv f.$Adiarnfd f.$Odiarnfc &&
#		git mv d.$Adiarnfd d.$Odiarnfc &&
#		git mv l.$Adiarnfd l.$Odiarnfc &&
#		git commit -m "mv Adiarnfd Odiarnfc"
#	

expecting success of 3910.14 'git checkout nfc': 
	rm f.$Odiarnfc &&
	git checkout f.$Odiarnfc

rm: cannot remove 'f.'$'\303\226': No such file or directory
not ok 14 - git checkout nfc
#	
#		rm f.$Odiarnfc &&
#		git checkout f.$Odiarnfc
#	

expecting success of 3910.15 'git checkout file nfd': 
	rm -f f.* &&
	git checkout f.$Odiarnfd

error: pathspec 'f.Ö' did not match any file(s) known to git
not ok 15 - git checkout file nfd
#	
#		rm -f f.* &&
#		git checkout f.$Odiarnfd
#	

expecting success of 3910.16 'git checkout link nfd': 
	rm l.* &&
	git checkout l.$Odiarnfd

error: pathspec 'l.Ö' did not match any file(s) known to git
not ok 16 - git checkout link nfd
#	
#		rm l.* &&
#		git checkout l.$Odiarnfd
#	

expecting success of 3910.17 'setup case mac2': 
	git checkout main &&
	git reset --hard &&
	git checkout -b mac_os_2

Switched to branch 'main'
HEAD is now at e53e959 rm x
Switched to a new branch 'mac_os_2'

expecting success of 3910.18 'commit file d2.Adiarnfd/f.Adiarnfd': 
	mkdir d2.$Adiarnfd &&
	echo d2.$Adiarnfd/f.$Adiarnfd >d2.$Adiarnfd/f.$Adiarnfd &&
	git add d2.$Adiarnfd/f.$Adiarnfd &&
	git commit -m "add d2.$Adiarnfd/f.$Adiarnfd" -- d2.$Adiarnfd/f.$Adiarnfd

ok 17 - setup case mac2
[mac_os_2 7c9f2db] add d2.Ä/f.Ä
 Author: A U Thor <author@example.com>
 1 file changed, 1 insertion(+)
 create mode 100644 "d2.A\314\210/f.A\314\210"

expecting success of 3910.19 'setup for long decomposed filename': 
	git checkout main &&
	git reset --hard &&
	git checkout -b mac_os_long_nfd_fn

ok 18 - commit file d2.Adiarnfd/f.Adiarnfd
Switched to branch 'main'
HEAD is now at e53e959 rm x
Switched to a new branch 'mac_os_long_nfd_fn'

expecting success of 3910.20 'Add long decomposed filename': 
	echo longd >$Alongd &&
	git add * &&
	git commit -m "Long filename"

ok 19 - setup for long decomposed filename
[mac_os_long_nfd_fn 022fe93] Long filename
 Author: A U Thor <author@example.com>
 3 files changed, 3 insertions(+)
 create mode 100644 "A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210A\314\210"
 create mode 100644 actual
 create mode 100644 expect

expecting success of 3910.21 'setup for long precomposed filename': 
	git checkout main &&
	git reset --hard &&
	git checkout -b mac_os_long_nfc_fn

ok 20 - Add long decomposed filename
Switched to branch 'main'
HEAD is now at e53e959 rm x
Switched to a new branch 'mac_os_long_nfc_fn'

expecting success of 3910.22 'Add long precomposed filename': 
	echo longc >$Alongc &&
	git add * &&
	git commit -m "Long filename"

ok 21 - setup for long precomposed filename
[mac_os_long_nfc_fn 1dd044c] Long filename
 Author: A U Thor <author@example.com>
 1 file changed, 1 insertion(+)
 create mode 100644 "\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206\303\206"

checking known breakage of 3910.23 'handle existing decomposed filenames': 
	echo content >"verbatim.$Adiarnfd" &&
	git -c core.precomposeunicode=false add "verbatim.$Adiarnfd" &&
	git commit -m "existing decomposed file" &&
	git ls-files --exclude-standard -o "verbatim*" >untracked &&
	test_must_be_empty untracked

ok 22 - Add long precomposed filename
[mac_os_long_nfc_fn 935384c] existing decomposed file
 Author: A U Thor <author@example.com>
 1 file changed, 1 insertion(+)
 create mode 100644 "verbatim.A\314\210"

expecting success of 3910.24 'unicode decomposed: git restore -p . ': 
	DIRNAMEPWD=dir.Odiarnfc &&
	DIRNAMEINREPO=dir.$Adiarnfc &&
	export DIRNAMEPWD DIRNAMEINREPO &&
	git init "$DIRNAMEPWD" &&
	(
		cd "$DIRNAMEPWD" &&
		mkdir "$DIRNAMEINREPO" &&
		cd "$DIRNAMEINREPO" &&
		echo "Initial" >file &&
		git add file &&
		echo "More stuff" >>file &&
		echo y | git restore -p .
	)

ok 23 - handle existing decomposed filenames # TODO known breakage vanished
Initialized empty Git repository in /build/git-2.37.3/t/trash directory.t3910-mac-os-precompose/dir.Odiarnfc/.git/
diff --git "a/dir.\303\204/file" "b/dir.\303\204/file"
index a77fa51..2a1fd39 100644
--- "a/dir.\303\204/file"
+++ "b/dir.\303\204/file"
@@ -1 +1,2 @@
 Initial
+More stuff
(1/1) Discard this hunk from worktree [y,n,q,a,d,e,?]? 

expecting success of 3910.25 'respect git config --global core.precomposeunicode': 
	git config --global core.precomposeunicode true &&
	rm -rf .git &&
	git init &&
	precomposeunicode=$(git config core.precomposeunicode) &&
	test "$precomposeunicode" = "true"

ok 24 - unicode decomposed: git restore -p . 
Initialized empty Git repository in /build/git-2.37.3/t/trash directory.t3910-mac-os-precompose/.git/

ok 25 - respect git config --global core.precomposeunicode
# 1 known breakage(s) vanished; please update test(s)
# failed 10 among remaining 24 test(s)
1..25
</pre>
</details>

winterqt commented:
At a glance, it looks like these all have to do with Unicode file names -- what are your zpool/dataset options for UTF-8 normalization, if any?

mweinelt commented:
```
tank                   normalization         formD                             -
[...]
```

winterqt commented:
I have no issues with merging this, but the commit message should probably be made more descriptive -- it doesn't fail universally with ZFS (I just checked on my machine without UTF-8 normalization enabled).

I'm going to take a look at this and (hopefully) send a patch upstream so this isn't needed for long, but no clue when I'll get a chance to do that.

mweinelt commented:
Rebased onto staging-next so I can use this in python-updates, where I currently do alot of work.

jonasnick commented:
Fwiw I tried setting up a fresh instance using ZFS following the [guide recommended in the wiki](https://openzfs.github.io/openzfs-docs/Getting%20Started/NixOS/Root%20on%20ZFS/1-preparation.html) (which sets normalization=formD), but immediately ran into issue https://github.com/NixOS/nixpkgs/issues/185882. This PR fixed it for me.
